### DISCUSSION TITLE
如何统计正文字数

### DISCUSSION THREAD

**Discussion by MARMOTatZJU** [TheNetAdmin/zjuthesis#108] [General]
请问有没有什么能够统计正文字数的好的实践？

**Comment by TheNetAdmin** [comment]
暂时没有通用的统计字数的方式，不过有以下几种方式可参考

1. 通过 texcount 工具统计，这个工具随 TeX Live 一起安装，但使用时不能很好地识别 `\input` 的内容（因为本模板 `\input` 的内容是由 `\Degree` 确定的）
2. 通过 pandoc 工具将 TeX 转换为其他格式（如纯文本或 Word）再统计
3. 将生成的 pdf 转换为其他格式再统计

**Comment by TheNetAdmin** [comment]
我在 2ad353c 中添加了统计字数用的 shell 脚本

该脚本调用的 `texcount` 工具是随 TeX Live 安装的，无需手动安装

**Comment by MARMOTatZJU** [comment]
@TheNetAdmin 十分感谢提供。这个可能更新到repo里会更好？或者repo下建一个docs目录，存放一些tutorial？

**Reply by TheNetAdmin** [reply]
你的意思是增加一些使用说明吧？我一会儿写一些使用说明，加到 README 里。

**Reply by MARMOTatZJU** [reply]
对的，感谢～

**Reply by MARMOTatZJU** [reply]
@TheNetAdmin 
实际使用的时候出现了以下信息：
```
!!! Encountered } without corresponding {. !!!
Deep recursion on subroutine "main::_parse_unit" at /Library/TeX/texbin/texcount line 2179.
```
我的文件能正常编译，所以应该不会存在花括号不匹配这样的syntax error。可能调用textcount的时候需要加一些参数？

**Reply by TheNetAdmin** [reply]
`tlmgr update --self --all`  升级所有宏包呢？以及你用的 TeX Live 是什么版本？

如果升级无效，开个 issue 我看下

**Reply by MARMOTatZJU** [reply]
版本为`Latexmk, John Collins, 29 September 2020. Version 4.70b`

已经升级，但是依然出现该问题。我开issues吧

**Comment by zy18051120439** [comment]
这里补一个自己使用过程中的简单说明。
在根目录zjuthesis下新建https://github.com/TheNetAdmin/zjuthesis/commit/2ad353cc002fb2868096f44b77b20d87fcc052af 中的sh文件。
1.新建文件夹，改名wc.sh，（wc.sh也就是word_count.sh的意思）右键打开方式，把上面链接里面的内容复制进去。
2.如果是windows，需要百度一下windows如何运行sh文件，根据https://jingyan.baidu.com/article/7082dc1cdc80a2e40a89bd22.html 百度经验安装个git。
3.在根目录zjuthesis下使用git bash here，输入sh wc.sh。
但是我个人使用了一下不是很好用，统计不到content的内容。
然后我发现这个脚本主要是执行
`#/bin/bash`
`texcount out/zjuthesis.wordcnt -inc`

这样一个命令。
然后根据路径，找到out/zjuthesis.wordcnt，右键打开方式记事本，发现是这样一些内容。

> \input{./body/graduate/post/ref.tex}
\input{./body/graduate/post/appendix.tex}

所以实际上这个文件夹里存放的就是想要统计的tex的路径，于是我将原来的bash直接删剩texcount那句话，然后在.wordcnt文件夹里手动添加想要统计字数的tex文件即可。
————————————————
即，
1.新建文件夹，改名wc.sh，（wc.sh也就是word_count.sh的意思，自己取，后缀是sh即可）右键打开方式，把链接https://github.com/TheNetAdmin/zjuthesis/commit/2ad353cc002fb2868096f44b77b20d87fcc052af  里面的内容复制进去。
2.如果是windows，需要百度一下windows如何运行sh文件，根据https://jingyan.baidu.com/article/7082dc1cdc80a2e40a89bd22.html 百度经验安装个git。
3.在根目录zjuthesis下使用git bash here，输入sh wc.sh。
4.若不好用，统计不全，可能可以尝试如下步骤，
  1)以记事本方式，打开wc.sh文件，将里面内容改为
    `#/bin/bash`
    `texcount out/zjuthesis.wordcnt -inc`
  2)以记事本方式打开out/zjuthesis.wordcnt，仿照文件夹里的格式：

> \input{./body/graduate/post/ref.tex}
\input{./body/graduate/post/appendix.tex}
\input{./body/graduate/content.tex}

仿照格式，手动输入自己想要统计字数的tex文件的路径以及文件名，然后回根目录运行sh wc.sh。

**Reply by TheNetAdmin** [reply]
`out/zjuthesis.fls` 是 TeXLive 编译时自动生成的，会包括正文所使用的各个文件。然后 `word_count.sh` 去掉其中不应当统计字数的文件（例如 pdf 文件），保存到 `zjuthesis.wordcnt` 并统计字数。

如果手动创建 `zjuthesis.wordcnt` 的话，记得要根据正文变化进行更新。

**Reply by zy18051120439** [reply]
谢谢提醒，谢谢大佬的模板~
